File: sdks/csharp/src/PipeWorker.cs
using System.Collections.Concurrent;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace RustPipe;

public delegate Task<object?> TaskHandler(PipeTask task);

public class PipeWorker : IAsyncDisposable
{
    private readonly string _url;
    private readonly string _workerId;
    private readonly Dictionary<string, TaskHandler> _handlers;
    private readonly int _maxConcurrency;
    private readonly TimeSpan _heartbeatInterval;
    private readonly SemaphoreSlim _semaphore;
    private readonly ClientWebSocket _ws = new();
    private readonly CancellationTokenSource _cts = new();
    private readonly DateTime _startTime = DateTime.UtcNow;
    private int _activeTasks;

    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    private PipeWorker(string url, string workerId, Dictionary<string, TaskHandler> handlers,
        int maxConcurrency, TimeSpan heartbeatInterval)
    {
        _url = url;
        _workerId = workerId;
        _handlers = handlers;
        _maxConcurrency = maxConcurrency;
        _heartbeatInterval = heartbeatInterval;
        _semaphore = new SemaphoreSlim(maxConcurrency);
    }

    public static Builder CreateBuilder(string url) => new(url);

    public async Task StartAsync()
    {
        await _ws.ConnectAsync(new Uri(_url), _cts.Token);
        await RegisterAsync();
        _ = HeartbeatLoopAsync();
        await ReceiveLoopAsync();
    }

    public async Task StopAsync()
    {
        _cts.Cancel();
        if (_ws.State == WebSocketState.Open)
            await _ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "shutdown", CancellationToken.None);
    }

    private async Task RegisterAsync()
    {
        var msg = new
        {
            type = "WorkerRegister",
            registration = new
            {
                workerId = _workerId,
                supportedTasks = _handlers.Keys.ToArray(),
                maxConcurrency = _maxConcurrency,
                language = "CSharp",
            }
        };
        await SendAsync(msg);
    }

    private async Task HeartbeatLoopAsync()
    {
        while (!_cts.IsCancellationRequested)
        {
            await Task.Delay(_heartbeatInterval, _cts.Token);
            var msg = new
            {
                type = "Heartbeat",
                payload = new
                {
                    workerId = _workerId,
                    activeTasks = _activeTasks,
                    capacity = _maxConcurrency,
                    uptimeSeconds = (int)(DateTime.UtcNow - _startTime).TotalSeconds,
                }
            };
            await SendAsync(msg);
        }
    }

    private async Task ReceiveLoopAsync()
    {
        var buffer = new byte[65536];
        while (!_cts.IsCancellationRequested && _ws.State == WebSocketState.Open)
        {
            var result = await _ws.ReceiveAsync(buffer, _cts.Token);
            if (result.MessageType == WebSocketMessageType.Close) break;

            var raw = Encoding.UTF8.GetString(buffer, 0, result.Count);
            var doc = JsonDocument.Parse(raw);
            var msgType = doc.RootElement.GetProperty("type").GetString();

            switch (msgType)
            {
                case "TaskDispatch":
                    var taskElem = doc.RootElement.GetProperty("task");
                    var task = new PipeTask(
                        taskElem.GetProperty("id").GetString()!,
                        taskElem.GetProperty("taskType").GetString()!,
                        taskElem.GetProperty("payload")
                    );
                    _ = ExecuteTaskAsync(task);
                    break;
                case "Shutdown":
                    await StopAsync();
                    break;
            }
        }
    }

    private async Task ExecuteTaskAsync(PipeTask task)
    {
        await _semaphore.WaitAsync(_cts.Token);
        Interlocked.Increment(ref _activeTasks);
        var start = DateTime.UtcNow;

        try
        {
            if (!_handlers.TryGetValue(task.TaskType, out var handler))
            {
                await SendResultAsync(task.Id, "Failed", null,
                    new { code = "UNKNOWN_TASK_TYPE", message = $"No handler for: {task.TaskType}", retryable = false }, 0);
                return;
            }

            var result = await handler(task);
            var duration = (int)(DateTime.UtcNow - start).TotalMilliseconds;
            await SendResultAsync(task.Id, "Completed", result, null, duration);
        }
        catch (Exception ex)
        {
            var duration = (int)(DateTime.UtcNow - start).TotalMilliseconds;
            await SendResultAsync(task.Id, "Failed", null,
                new { code = "HANDLER_ERROR", message = ex.Message, retryable = true }, duration);
        }
        finally
        {
            Interlocked.Decrement(ref _activeTasks);
            _semaphore.Release();
        }
    }

    private async Task SendResultAsync(string taskId, string status, object? payload, object? error, int durationMs)
    {
        var msg = new
        {
            type = "TaskResult",
            result = new
            {
                taskId,
                status,
                payload,
                error,
                durationMs,
                workerId = _workerId,
            }
        };
        await SendAsync(msg);
    }

    private async Task SendAsync(object message)
    {
        var json = JsonSerializer.Serialize(message, JsonOpts);
        var bytes = Encoding.UTF8.GetBytes(json);
        await _ws.SendAsync(bytes, WebSocketMessageType.Text, true, _cts.Token);
    }

    public async ValueTask DisposeAsync()
    {
        await StopAsync();
        _ws.Dispose();
        _cts.Dispose();
    }

    public class Builder
    {
        private readonly string _url;
        private string? _workerId;
        private readonly Dictionary<string, TaskHandler> _handlers = new();
        private int _maxConcurrency = 10;
        private TimeSpan _heartbeatInterval = TimeSpan.FromSeconds(5);

        internal Builder(string url) { _url = url; }

        public Builder WorkerId(string id) { _workerId = id; return this; }
        public Builder MaxConcurrency(int n) { _maxConcurrency = n; return this; }
        public Builder HeartbeatInterval(TimeSpan interval) { _heartbeatInterval = interval; return this; }
        public Builder Handle(string taskType, TaskHandler handler) { _handlers[taskType] = handler; return this; }

        public PipeWorker Build()
        {
            var workerId = _workerId ?? $"worker-csharp-{Guid.NewGuid().ToString()[..8]}";
            return new PipeWorker(_url, workerId, _handlers, _maxConcurrency, _heartbeatInterval);
        }
    }
}

public record PipeTask(string Id, string TaskType, JsonElement Payload);

File: sdks/php/src/PipeWorker.php
<?php

namespace RustPipe;

use WebSocket\Client;
use Ramsey\Uuid\Uuid;

class PipeWorker
{
    private string $url;
    private string $workerId;
    private array $handlers;
    private int $maxConcurrency;
    private float $heartbeatInterval;
    private int $activeTasks = 0;
    private float $startTime;
    private ?Client $ws = null;
    private bool $running = false;

    public function __construct(
        string $url,
        array $handlers,
        ?string $workerId = null,
        int $maxConcurrency = 10,
        float $heartbeatInterval = 5.0,
    ) {
        $this->url = $url;
        $this->handlers = $handlers;
        $this->workerId = $workerId ?? 'worker-php-' . substr(Uuid::uuid4()->toString(), 0, 8);
        $this->maxConcurrency = $maxConcurrency;
        $this->heartbeatInterval = $heartbeatInterval;
        $this->startTime = microtime(true);
    }

    public function start(): void
    {
        $this->running = true;
        $this->connect();
    }

    public function stop(): void
    {
        $this->running = false;
        $this->ws?->close();
    }

    private function connect(): void
    {
        $this->ws = new Client($this->url);
        $this->register();

        $lastHeartbeat = time();

        while ($this->running) {
            try {
                $raw = $this->ws->receive();
                if ($raw) {
                    $this->handleMessage($raw);
                }
            } catch (\Exception $e) {
                if ($this->running) {
                    sleep(1);
                    $this->ws = new Client($this->url);
                    $this->register();
                }
            }

            if (time() - $lastHeartbeat >= $this->heartbeatInterval) {
                $this->sendHeartbeat();
                $lastHeartbeat = time();
            }
        }
    }

    private function register(): void
    {
        $this->send([
            'type' => 'WorkerRegister',
            'registration' => [
                'workerId' => $this->workerId,
                'supportedTasks' => array_keys($this->handlers),
                'maxConcurrency' => $this->maxConcurrency,
                'language' => 'PHP',
            ],
        ]);
    }

    private function sendHeartbeat(): void
    {
        $this->send([
            'type' => 'Heartbeat',
            'payload' => [
                'workerId' => $this->workerId,
                'activeTasks' => $this->activeTasks,
                'capacity' => $this->maxConcurrency,
                'uptimeSeconds' => (int)(microtime(true) - $this->startTime),
            ],
        ]);
    }

    private function handleMessage(string $raw): void
    {
        $msg = json_decode($raw, true);
        if (!$msg || !isset($msg['type'])) return;

        match ($msg['type']) {
            'TaskDispatch' => $this->executeTask($msg['task']),
            'Shutdown' => $this->stop(),
            default => null,
        };
    }

    private function executeTask(array $task): void
    {
        $this->activeTasks++;
        $start = microtime(true);
        $taskId = $task['id'];
        $taskType = $task['taskType'];

        try {
            $handler = $this->handlers[$taskType] ?? null;
            if (!$handler) {
                $this->sendResult($taskId, 'Failed', null, [
                    'code' => 'UNKNOWN_TASK_TYPE',
                    'message' => "No handler for: $taskType",
                    'retryable' => false,
                ], 0);
                return;
            }

            $result = $handler($task);
            $duration = (int)((microtime(true) - $start) * 1000);
            $this->sendResult($taskId, 'Completed', $result, null, $duration);
        } catch (\Exception $e) {
            $duration = (int)((microtime(true) - $start) * 1000);
            $this->sendResult($taskId, 'Failed', null, [
                'code' => 'HANDLER_ERROR',
                'message' => $e->getMessage(),
                'retryable' => true,
            ], $duration);
        } finally {
            $this->activeTasks--;
        }
    }

    private function sendResult(string $taskId, string $status, mixed $payload, ?array $error, int $durationMs): void
    {
        $result = [
            'taskId' => $taskId,
            'status' => $status,
            'durationMs' => $durationMs,
            'workerId' => $this->workerId,
        ];
        if ($payload !== null) $result['payload'] = $payload;
        if ($error !== null) $result['error'] = $error;

        $this->send(['type' => 'TaskResult', 'result' => $result]);
    }

    private function send(array $message): void
    {
        $this->ws?->text(json_encode($message));
    }
}
